src/core/packages/saved-objects/api-server-internal/src/lib/apis/internals/change_object_access_control.ts

```diff
@@ -0,0 +1,369 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+import type { estypes } from '@elastic/elasticsearch';
+import { isNotFoundFromUnsupportedServer } from '@kbn/core-elasticsearch-server-internal';
+import {
+  type ISavedObjectTypeRegistry,
+  type ISavedObjectsSecurityExtension,
+  type ISavedObjectsSerializer,
+  SavedObjectsErrorHelpers,
+  type SavedObjectsRawDoc,
+  type SavedObjectsRawDocSource,
+} from '@kbn/core-saved-objects-server';
+import type {
+  SavedObjectsChangeAccessControlResponse,
+  SavedObjectsChangeAccessControlObject,
+  SavedObjectsChangeAccessControlOptions,
+  SavedObjectsChangeAccessModeOptions,
+  SavedObjectsChangeOwnershipOptions,
+} from '@kbn/core-saved-objects-api-server';
+
+import {
+  getBulkOperationError,
+  getExpectedVersionProperties,
+  rawDocExistsInNamespace,
+  type Either,
+  isLeft,
+  isRight,
+  left,
+  right,
+} from '../utils';
+import type { ApiExecutionContext } from '../types';
+import { type GetBulkOperationErrorRawResponse, isMgetError } from '../utils/internal_utils';
+
+export type ChangeAccessControlActionType = 'changeOwnership' | 'changeAccessMode';
+
+export interface ChangeAccessControlParams {
+  registry: ISavedObjectTypeRegistry;
+  allowedTypes: string[];
+  client: ApiExecutionContext['client'];
+  serializer: ISavedObjectsSerializer;
+  getIndexForType: (type: string) => string;
+  objects: SavedObjectsChangeAccessControlObject[];
+  options: SavedObjectsChangeAccessControlOptions;
+  securityExtension?: ISavedObjectsSecurityExtension;
+  actionType: ChangeAccessControlActionType;
+  currentUserProfileUid: string;
+}
+
+// Regular expression to validate user profile IDs as generated by Elasticsearch
+// User profile IDs are expected to be in the format "u_<principal>_<version>"
+const USER_PROFILE_REGEX = /^u_.+_.+$/;
```

Comment:
This regexp seems incorrect. It will match strings like:

```ts
/^u_.+_.+$/.test('u_asd_asd_asd');
// => true
```

Unless `asd_asd` is a valid principle name?

Perhaps we can avoid regexp and write logic like: `const [u, principal, version, notAllowed] = string.split('_')`